File: backend/app/schemas/personnel.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class DiplomeBase(BaseModel):
    libelle: str = Field(..., min_length=2, max_length=200)
    etablissement: str | None = None
    annee_obtention: int | None = Field(default=None, ge=1950, le=2100)
    niveau: str | None = None


class DiplomeCreate(DiplomeBase):
    pass


class DiplomeUpdate(BaseModel):
    libelle: str | None = Field(default=None, min_length=2, max_length=200)
    etablissement: str | None = None
    annee_obtention: int | None = Field(default=None, ge=1950, le=2100)
    niveau: str | None = None


class DiplomeResponse(DiplomeBase):
    id: UUID
    personnel_id: UUID

    model_config = {"from_attributes": True}


class ContratBase(BaseModel):
    type_contrat: str = Field(..., pattern=r"^(cdi|cdd|vacataire)$")
    date_debut: date
    date_fin: date | None = None
    salaire_mensuel: Decimal | None = Field(default=None, ge=0)
    statut: str = Field(default="actif", pattern=r"^(actif|termine)$")


class ContratCreate(ContratBase):
    pass


class ContratUpdate(BaseModel):
    type_contrat: str | None = Field(default=None, pattern=r"^(cdi|cdd|vacataire)$")
    date_debut: date | None = None
    date_fin: date | None = None
    salaire_mensuel: Decimal | None = Field(default=None, ge=0)
    statut: str | None = Field(default=None, pattern=r"^(actif|termine)$")


class ContratResponse(ContratBase):
    id: UUID
    personnel_id: UUID

    model_config = {"from_attributes": True}


class ClasseBrief(BaseModel):
    id: UUID
    nom: str

    model_config = {"from_attributes": True}


class MatiereBrief(BaseModel):
    id: UUID
    code: str
    libelle: str

    model_config = {"from_attributes": True}


class AffectationCreate(BaseModel):
    classe_id: UUID
    matiere_id: UUID
    annee_scolaire_id: UUID | None = None


class AffectationResponse(BaseModel):
    id: UUID
    personnel_id: UUID
    classe_id: UUID
    matiere_id: UUID
    annee_scolaire_id: UUID
    classe: ClasseBrief | None = None
    matiere: MatiereBrief | None = None

    model_config = {"from_attributes": True}


class TitulaireRequest(BaseModel):
    classe_id: UUID


class CongeBase(BaseModel):
    type: str = Field(..., pattern=r"^(conge|maladie|permission|absence)$")
    date_debut: date
    date_fin: date
    motif: str | None = None
    statut: str = Field(default="demande", pattern=r"^(demande|approuve|refuse)$")


class CongeCreate(CongeBase):
    pass


class CongeUpdate(BaseModel):
    type: str | None = Field(default=None, pattern=r"^(conge|maladie|permission|absence)$")
    date_debut: date | None = None
    date_fin: date | None = None
    motif: str | None = None
    statut: str | None = Field(default=None, pattern=r"^(demande|approuve|refuse)$")


class CongeResponse(CongeBase):
    id: UUID
    personnel_id: UUID

    model_config = {"from_attributes": True}


class PersonnelCreate(BaseModel):
    nom: str = Field(..., min_length=1, max_length=100)
    prenoms: str = Field(..., min_length=1, max_length=150)
    sexe: str = Field(..., pattern=r"^(M|F)$")
    date_naissance: date | None = None
    telephone: str = Field(..., min_length=6, max_length=20)
    email: str | None = None
    adresse: str | None = None
    categorie: str = Field(..., pattern=r"^(enseignant|non_enseignant)$")
    fonction: str | None = None
    specialite: str | None = None
    date_embauche: date | None = None
    user_id: UUID | None = None


class PersonnelUpdate(BaseModel):
    nom: str | None = Field(default=None, min_length=1, max_length=100)
    prenoms: str | None = Field(default=None, min_length=1, max_length=150)
    sexe: str | None = Field(default=None, pattern=r"^(M|F)$")
    date_naissance: date | None = None
    telephone: str | None = Field(default=None, min_length=6, max_length=20)
    email: str | None = None
    adresse: str | None = None
    fonction: str | None = None
    specialite: str | None = None
    date_embauche: date | None = None
    statut: str | None = Field(default=None, pattern=r"^(actif|inactif)$")
    user_id: UUID | None = None


class PersonnelListItem(BaseModel):
    id: UUID
    matricule: str
    nom: str
    prenoms: str
    sexe: str
    telephone: str
    email: str | None
    categorie: str
    fonction: str | None
    specialite: str | None
    statut: str

    model_config = {"from_attributes": True}


class PersonnelResponse(BaseModel):
    id: UUID
    matricule: str
    nom: str
    prenoms: str
    sexe: str
    date_naissance: date | None
    telephone: str
    email: str | None
    adresse: str | None
    categorie: str
    fonction: str | None
    specialite: str | None
    date_embauche: date | None
    statut: str
    user_id: UUID | None
    created_at: datetime
    diplomes: list[DiplomeResponse] = []
    contrats: list[ContratResponse] = []
    affectations: list[AffectationResponse] = []
    conges: list[CongeResponse] = []
    classes_titulaire: list[ClasseBrief] = []

    model_config = {"from_attributes": True}


class PersonnelListResponse(BaseModel):
    items: list[PersonnelListItem]
    total: int

File: frontend/src/types/personnel.ts
export interface PersonnelListItem {
  id: string;
  matricule: string;
  nom: string;
  prenoms: string;
  sexe: string;
  telephone: string;
  email: string | null;
  categorie: string;
  fonction: string | null;
  specialite: string | null;
  statut: string;
}

export interface PersonnelListResponse {
  items: PersonnelListItem[];
  total: number;
}

export interface Diplome {
  id: string;
  personnel_id: string;
  libelle: string;
  etablissement: string | null;
  annee_obtention: number | null;
  niveau: string | null;
}

export interface Contrat {
  id: string;
  personnel_id: string;
  type_contrat: string;
  date_debut: string;
  date_fin: string | null;
  salaire_mensuel: number | null;
  statut: string;
}

export interface Affectation {
  id: string;
  personnel_id: string;
  classe_id: string;
  matiere_id: string;
  annee_scolaire_id: string;
  classe?: { id: string; nom: string };
  matiere?: { id: string; code: string; libelle: string };
}

export interface Conge {
  id: string;
  personnel_id: string;
  type: string;
  date_debut: string;
  date_fin: string;
  motif: string | null;
  statut: string;
}

export interface Personnel {
  id: string;
  matricule: string;
  nom: string;
  prenoms: string;
  sexe: string;
  date_naissance: string | null;
  telephone: string;
  email: string | null;
  adresse: string | null;
  categorie: string;
  fonction: string | null;
  specialite: string | null;
  date_embauche: string | null;
  statut: string;
  user_id: string | null;
  created_at: string;
  diplomes: Diplome[];
  contrats: Contrat[];
  affectations: Affectation[];
  conges: Conge[];
  classes_titulaire: { id: string; nom: string }[];
}

export interface CreatePersonnelRequest {
  nom: string;
  prenoms: string;
  sexe: string;
  telephone: string;
  categorie: string;
  date_naissance?: string;
  email?: string;
  adresse?: string;
  fonction?: string;
  specialite?: string;
  date_embauche?: string;
}
